### PR TITLE
Save .command.log for k8s tasks

### LINKED ISSUE #2364: Task .command.log not saved by k8s executor
## Bug report 

The k8s executor doesn't save `.command.log` for each task. This matters because output from `beforeScript` and `afterScript` are not saved to `.command.out` or `.command.err`, so I can't capture outputs from these directives when using kubernetes. Task logs can be accessed via `kubectl logs <pod>` but pods are deleted upon completion.

I see that this functionality was added recently in the case of pod failures (#699), I think this behaviour should be extended to pod completions as well.

Alternatively we could change `beforeScript` and `afterScript` to also pipe into `.command.out` / `.command.err` but I think that would be more complicated.

### Environment 

* Nextflow version: 21.08-edge
* Java version: 1.8
* Operating system: Linux

### PR DESCRIPTION
**Pull request by bentsherman**
Resolves #2364 

This PR updates the K8s task handler to always save `.command.log`. I also removed some code that saved the log for failed pods in a more roundabout way.

However, although the build succeeds and the relevant tests pass, I haven't gotten it to work in my local environment. The task pod command is still:
```yaml
    command:
    - '/bin/bash'
    - '-ue'
    - '.command.run'
```

when it should be:
```yaml
    command:
    - '/bin/bash'
    - '-ue'
    - 'bash .command.run 2>&1 | tee .command.log'
```

Am I missing something?

### CODE REVIEW

**Comment by pditommaso** [comment]
Has this been tested?

File: modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy [hunk 1]
```diff
@@ -161,7 +160,7 @@ class K8sTaskHandler extends TaskHandler {
     protected Map newSubmitRequest0(TaskRun task, String imageName) {
 
         final fixOwnership = builder.fixOwnership()
-        final cmd = new ArrayList(new ArrayList(BashWrapperBuilder.BASH)) << TaskRun.CMD_RUN
+        final cmd = new ArrayList(BashWrapperBuilder.BASH) << "bash ${TaskRun.CMD_RUN} 2>&1 | tee ${TaskRun.CMD_LOG}"
```
[the remark this answers is not in the document]
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy:163]: Just found this 😆 

https://twitter.com/aahoogendoorn/status/1513550482752643082

[the remark this answers is not in the document]
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy:163]: If you have a local K8s stack should be quite easy to test it, using this config snippet 

```
process {
   executor = 'k8s'
}

k8s {
  autoMountHostPaths = true
  debug {
    yaml = true
  } 
}
```

and then run nextflow using `./launch.sh run <something> -c <the config above>

**Comment by pditommaso** [comment]
Do you think there's a real need for this log? or it's just for consistency with other executors?

**Comment by bentsherman** [comment]
There are cases where things get printed to `.command.log` but not `.command.out` / `.command.err`. I ran into these cases during my PhD work. But it would also make the nextflow / k8s experience more consistent with other executors.

I think these changes are correct, but it wasn't working during the full test because `kuberun` wasn't using my development version of Nextflow on the k8s cluster. I assume the `autoMountHostPaths` allows you to bypass that on a local k8s cluster so I will try that.

File: modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy]: You need to use `run` not `kuberun`

**Comment by bentsherman** [comment]
Okay, I figured out how to test with k8s like you showed, but I hit some weird errors. So I backed up to `master`, tried again, and got this error:
```
 ./launch.sh -c k8s.config run hello
N E X T F L O W  ~  version 22.05.0-SNAPSHOT
Launching `https://github.com/nextflow-io/hello` [mighty_goldwasser] DSL2 - revision: 4eab81bd42 [master]
executor >  k8s (4)
executor >  k8s (4)
[77/d802dd] process > sayHello (1) [ 75%] 3 of 4, failed: 3
Error executing process > 'sayHello (3)'

Caused by:
  Process `sayHello (3)` terminated for an unknown reason -- Likely it has been terminated by the external system

Command executed:

  echo 'Hello world!'

Command exit status:
  -

Command output:
  (empty)

Command wrapper:
  /bin/bash: .command.run: No such file or directory

Work dir:
  /home/bent/workspace/nextflow-io_nextflow/work/eb/7880bd01e64a40002f55f7ade78c6b
```

However, I don't see anything wrong with the pod yaml:
```
apiVersion: v1
kind: Pod
metadata:
  name: nf-eb7880bd01e64a40002f55f7ade78c6b
  namespace: tower-nf
  labels: {app: nextflow, taskName: sayHello_3, sessionId: uuid-0b909bb8-de05-49e2-99b1-d4d07b3aad1c,
    processName: sayHello, runName: mighty_goldwasser}
spec:
  restartPolicy: Never
  containers:
  - name: nf-eb7880bd01e64a40002f55f7ade78c6b
    image: quay.io/nextflow/bash
    command: [/bin/bash, -ue, /home/bent/workspace/nextflow-io_nextflow/work/eb/7880bd01e64a40002f55f7ade78c6b/.command.run]
    resources:
      requests: {cpu: 1}
      limits: {cpu: 1}
    volumeMounts:
    - {name: vol-1, mountPath: /home/bent/workspace/nextflow-io_nextflow/work/eb/7880bd01e64a40002f55f7ade78c6b}
  volumes:
  - name: vol-1
    hostPath: {path: /home/bent/workspace/nextflow-io_nextflow/work/eb/7880bd01e64a40002f55f7ade78c6b}
```

It says it can't find `.command.run` but the work dir is mounted to the pod and `.command.run` is there.

**Comment by pditommaso** [comment]
ok, found the reasons, the pipe concatenation should be executed as a subshell. But still not sure, not see a big value to add this log because this is running within the container. 

it only helps when the .command.run crashes .. (that the reason why you want to add :))

**Comment by bentsherman** [comment]
I believe the original motivation was that `beforeScript` and `afterScript` are only saved to `.command.log`. See the related issue for details.

Even if it's a small improvement, why not do it? The awsbatch executor does the same thing for a similar reason: https://github.com/nextflow-io/nextflow/blob/7fa4c86c353c962c6f6840ce441f4dbd13aa605d/plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsBatchTaskHandler.groovy#L600-L611

**Comment by bentsherman** [comment]
Also, thanks for the fix. I thought it was already executing in a shell but I forgot about the `-c` option.

**Comment by pditommaso** [comment]
> Even if it's a small improvement, why not do it?

Well, mainly to prevent the creation of files, but I agree it can be useful to debug the launcher script and it's consistent with the existing pattern. 

Can you please have a look to the failing tests and then I'll merge it. tx!

File: modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy [hunk 2]
```diff
@@ -379,28 +381,6 @@ class K8sTaskHandler extends TaskHandler {
...
-                    ? client.jobLog(podName)
-                    : client.podLog(podName)
-            Files.copy(stream, task.workDir.resolve(TaskRun.CMD_LOG))
-        }
-        catch( Exception e ) {
-            log.warn "Failed to copy log for ${resourceType.lower()} $podName", e
-        }
-    }
-
     protected int readExitFile() {
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy:384]: Why this was removed?

Comment (bentsherman) [reply modules/nextflow/src/main/groovy/nextflow/k8s/K8sTaskHandler.groovy:384]: This method would save the command log for failed pods, so it's no longer needed.

**Comment by bentsherman** [comment]
@pditommaso This PR is old so I updated it and cleaned it up. It is ready to merge whenever you want.

**Comment by pditommaso** [comment]
I believe fetching the log over the stream it's better because it allows capturing the error output even when the `.command.log` cannot be accessed (think Fusion enable to mount the bucket for example). 

Likely the best would be to implement both approaches (.command.log and fallback to the stream if missing), but I think it's overkill without a real need. 

Closing without merging. We can reconsider it in the future.

**Comment by bentsherman** [comment]
Fetching the log in this manner does not work when the pod is deleted immediately, for example:

```
WARN: Failed to copy log for job nf-b29b6c384d74ee3b41844239686e8a6c
```

**Comment by pditommaso** [comment]
The pod should not be deleted

**Pull request closed without merging**